### DISCUSSION TITLE
Color Segmentation

### DISCUSSION THREAD

**Discussion by ojwebb** [ImageMagick/ImageMagick#4909] [Digital Image Processing]
ImageMagick 7.1.0-26 Q16 on a MAC
Looking for a way to segment a logo into individual colors while including edge transitional colors based on the closest local contiguous solid color. Wondering if the can be identify the contiguous unique colors>remap or quantize>then segment based on that color. Thanks Guys

**Comment by snibgo** [comment]
Many methods can be used for this.

Perhaps the most obvious method assumes there are no more than 255 solid colours. Reduce to 255 colours, then examine the histogram of colours, ie the count of each different colour. The solid colours will (probably) have a large count, and the anti-aliased colours will have a small count. So now you know which solid colours have been used, and can remap to just those colours.

Other possibilities: use "-kmeans" or other colour-reduction operation with gradually increasing number of colours. When the difference between the result and the original input is low, you can assume the result is just the solid colours. The "Jump Method" can help with the processing to identify when the difference "jumps" down. See ['Jump' method](http://im.snibgo.com/jumpmeth.htm), which gives a Windows BAT script.

EDIT: And, as you suggest, look for "contiguous unique colors". I would do this with gradients. Where the gradient is zero, that pixel is the same colour as its neighbours. Make a mask of these zero-gradient pixels, and that gives a list of the solid colours, so you can remap to them.
